libcontainer/nsenter/nsexec.c

```diff
@@ -1053,6 +1063,29 @@ void nsexec(void)
 					bail("failed to sync with parent: SYNC_MOUNTSOURCES_ACK: got %u", s);
 			}
 
+			/*
+			 * Set temporary single CPU affinity before cgroup cpuset transition,
+			 * this handles a corner case when joining a container having all
+			 * the processes running exclusively on isolated CPU cores to force
+			 * the kernel to schedule runc process on the first CPU core within the
+			 * cgroups cpuset. The introduction of the kernel commit
+			 * 46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76 has affected this deterministic
+			 * scheduling behavior by distributing tasks across CPU cores within the
+			 * cgroups cpuset. Some intensive real-time application are relying on this
+			 * deterministic behavior and use the first CPU core to run a slow thread
+			 * while other CPU cores are fully used by real-time threads with SCHED_FIFO
+			 * policy. Such applications prevent runc process from joining a container
+			 * when the runc process is randomly scheduled on a CPU core owned by a
+			 * real-time thread.
+			 */
+			if ((int32_t) config.cpu_affinity >= 0) {
+				cpu_set_t set;
+				CPU_ZERO(&set);
+				CPU_SET(config.cpu_affinity, &set);
+				if (sched_setaffinity(0, sizeof(set), &set) == -1)
```

Comment:
@cclerget I have tested your solution in my host. There is a big problem, the cpu affinity has the higher priority then `cpuset.cpus`. All tasks in the container will be scheduled to this tempporary single CPU.
I think you should remove the cpu affinity before the container started.

Reply:
@lifubang When the process is joining an existing container, the CPU affinity mask is reset to the cgroup cpuset, `taskset -c -p $$` returns the cpu affinity mask corresponding to cpuset.

What you've observed is expected, as the scheduler do not schedule tasks pinned to isolated CPUs, that's why this [patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) is problematic.
Since the scheduler doesn't manage tasks on isolated CPUs, everything is executed on the temporary CPU, my changes just restore the kernel behavior prior to kernel patch above.
You can test without my changes, you will also observe that all processes runs on the same isolated CPU, the difference is that the isolated CPU will randomly change each time you are entering in the container, but subsequent processes still runs on the same isolated CPU.

Removing the cpu affinity before executing container process would defeat the purpose of this patch, as the kernel could pin it randomly on a "locked" isolated core, this is the last case from [patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) description:

```
The cases that this mainly affects are:
 - modifying cpuset.cpus
 - when tasks join a cpuset
 - when modifying a task's affinity via sched_setaffinity(2)
```

Reply:
> What you've observed is expected,

I think it is not expected.
For example:
```bash
root@acmcoder:/home/acmcoder# cat /sys/devices/system/cpu/isolated
0-1,3
```
ContainerA: cpuset.cpus: "1,3"
When the ContainerA started, all the tasks in this container will be scheduled to `cpu1`, no tasks in `cpu3`.

Reply:
> You can test without my changes, you will also observe that all processes runs on the same isolated CPU,

No, I have tested it. With the same configuration as above, both `cpu1` and `cpu3` have  the same level load.

Reply:
> ContainerA: cpuset.cpus: "1,3"
When the ContainerA started, all the tasks in this container will be scheduled to cpu1, no tasks in cpu3.

Just to be sure, the patch is affecting `runc exec` not `runc run` (`initSetns` only https://github.com/opencontainers/runc/pull/3923/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2251)

Let me clarify and describe what I did for testing, I have a VM with 8 CPUs with `isolcpus=0-3`.

I'm using an OCI config containing, the container just call sleep for some time :

```
	"linux": {
		"cgroupsPath": "/repro",
		"resources": {
			"cpu": {
				"cpus": "0-3",
				"mems": "0"
			},
			"devices": [
				{
					"allow": false,
					"access": "rwm"
				}
			]
		},
		"namespaces": [
			{
				"type": "mount"
			}
		]
	}
```

In a first terminal I'm starting the container with:

```
sudo runc run --pid-file /tmp/repro.pid --bundle /tmp/bundle repro
```

Then in a second terminal I'm spawing another process to test the patch :

```
sudo runc exec repro sh -c "for i \$(seq 1 100); do cat /proc/self/stat|awk '{print $39}'; done"
```

I did that with and without my runc changes and with `isolcpus=0-3` for both, all iterations will show the exact same CPU between 0-3 for the unmodified runc (this is expected, this is the kernel [patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) in action):

```
sudo runc exec repro sh -c "for i \$(seq 1 100); do cat /proc/self/stat|cut -d ' ' -f 39; done"
2
2
2
2
2
2
....
```

While modified runc will show 0 for all (expected result), my patch spawn the entering process on the first CPU of the `cpuset.cpus`.


```
sudo runc exec repro sh -c "for i \$(seq 1 100); do cat /proc/self/stat|cut -d ' ' -f 39; done"
2
2
2
2
2
2
....
```
Now if you run the same without `isolcpus` set in the VM, you will notice that all iterations are showing different CPU for each iterations.

Reply:
Your clarify is clear, I have understood all your opinions, thanks.

Reply:
Cool ! I also pushed changes related to https://github.com/opencontainers/runc/pull/3923#discussion_r1264319356